src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs

```diff
@@ -0,0 +1,293 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
+using System.Diagnostics;
+using System.Net;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Net.WebSockets
+{
+    public sealed class ClientWebSocket : WebSocket
+    {
+        private enum InternalState
+        {
+            Created = 0,
+            Connecting = 1,
+            Connected = 2,
+            Disposed = 3
+        }
+
+        private const string WebSocketAvailableApiCheck = "WinHttpWebSocketCompleteUpgrade";
+
+        private readonly ClientWebSocketOptions _options;
+        private WinHttpWebSocket _innerWebSocket;
+        private readonly CancellationTokenSource _cts;
+
+        // NOTE: this is really an InternalState value, but Interlocked doesn't support
+        //       operations on values of enum types.
+        private int _state;
+
+        public ClientWebSocket()
+        {
+            if (Logging.On)
+            {
+                Logging.Enter(Logging.WebSockets, this, ".ctor", null);
+            }
+
+            CheckPlatformSupport();
```

Comment:
It looks like we're doing some non-trivial checks inside of this, loading a library, getting a function address, etc.  Should the result of this be cached rather than doing it each time one of these ClientWebSockets is instantiated?  Or am I wrong and it's fairly cheap?  Or maybe typical use cases don't instantiate many of these?